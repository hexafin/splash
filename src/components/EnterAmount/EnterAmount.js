import React, { Component } from "react";
import {
	Animated,
	Dimensions,
	TouchableWithoutFeedback,
	View,
	StyleSheet,
	Text,
	Image,
	Easing,
} from "react-native";
import { defaults, icons } from "../../lib/styles";
import { colors } from "../../lib/colors";
import { isIphoneX } from "react-native-iphone-x-helper";
import ReactNativeHapticFeedback from "react-native-haptic-feedback";
import Keypad from "../universal/Keypad";
import CloseButton from "../universal/CloseButton";
import NextButton from "../universal/NextButton";
import CurrencySwitcherLight from "../universal/CurrencySwitcherLight";
import NavigatorService from "../../redux/navigator";
import { cryptoTitleDict } from "../../lib/cryptos"

const SCREEN_WIDTH = Dimensions.get("window").width;
const SCREEN_HEIGHT = Dimensions.get("window").height;

class EnterAmount extends Component {
	constructor(props) {
		super(props)
		this.state = {
			amount: "",
			activeCurrency: props.activeCurrency,
		}
	}

	componentWillMount() {
		this.amountScale = new Animated.Value(1)
	}

	componentWillReceiveProps(nextProps) {
		if (nextProps.activeCurrency != this.state.activeCurrency) {
			this.setState({activeCurrency: nextProps.activeCurrency})
		}
	}

	shouldComponentUpdate(nextProps, nextState) {

		if (nextState.amount.length > 5) {
			Animated.timing(this.amountScale, {
				toValue: 1 - (nextState.amount.length-5)*0.1,
				duration: 100,
			}).start()
		}
		else if (nextState.amount.length <= 5 && this.state.amount.length > 5) {
			Animated.timing(this.amountScale, {
				toValue: 1,
				duration: 100,
			}).start()
		}

		if (nextState.activeCurrency != this.state.activeCurrency) {
			// update amount with exchange rate
			let newAmount
			if (this.state.amount != "") {
				newAmount = nextState.activeCurrency == "USD" 
					? (parseFloat(this.state.amount) * this.props.exchangeRates.BTC.USD).toFixed(2).slice(0, 7)
					: (parseFloat(this.state.amount) / this.props.exchangeRates.BTC.USD).toFixed(5).slice(0, 7)
			}
			else {
				newAmount = ""
			}
			this.setState({amount: newAmount.toString() })
			return true
		}
		else if (nextState.amount != this.state.amount) {
			return true
		}
		else {
			return false
		}
	}

	render() {
		const btcBalance = this.props.balance.BTC
		const balance = {
			BTC: btcBalance.toFixed(5),
			USD: (btcBalance * (this.props.exchangeRates.BTC ? this.props.exchangeRates.BTC.USD : 0)).toFixed(2)
		}
		const amountOverBalance = (this.state.activeCurrency == "USD" 
			? balance.USD < parseFloat(this.state.amount)
			: balance.BTC < parseFloat(this.state.amount))
		return (
			<View style={styles.wrapper}>

				<View style={styles.header}>
					<Text style={styles.title}>Sending Bitcoin</Text>
				</View>

				<View style={styles.bodyWrapper}>
					<CurrencySwitcherLight
						fiat="USD" 
						crypto="BTC" 
						textColor={colors.primary} 
						switcherColor={"purple"}
						textSize={16}
						onSwitch={currency => {
							this.setState({activeCurrency: currency})
						}}
						style={{position: "absolute", left: 20, top: 40}}/>

					<View style={styles.amountWrapper} pointerEvents="none">
						<Animated.View style={[styles.amountTextWrapper, {
							transform: [{scale: this.amountScale}]
						}]}>
							{this.state.activeCurrency == "USD" && 
								<Text style={[styles.amountSuffix, {
									fontSize: 34, 
									paddingRight: 3,
									color: amountOverBalance ? colors.red : colors.primary,
								}]}>
									$
								</Text>}

							<Text style={[styles.amount, {
								opacity: this.state.amount == "" ? 0.8 : 1.0,
								color: amountOverBalance ? colors.red : colors.primary,
							}]}>
								{this.state.amount == "" ? 0 : this.state.amount}
							</Text>

							{this.state.activeCurrency == "BTC" && 
								<Text style={[styles.amountSuffix, {
									fontSize: 30, 
									paddingLeft: 5,
									color: amountOverBalance ? colors.red : colors.primary,
								}]}>BTC</Text>}
						</Animated.View>
						<Text style={[styles.balance, {
							color: amountOverBalance ? colors.red : colors.gray,
						}]}>
							Balance: {this.state.activeCurrency == "USD" && "$"}{balance[this.state.activeCurrency]} 
							{this.state.activeCurrency == "BTC" && " BTC"}
						</Text>
					</View>
				</View>

				<Keypad
					primaryColor={"#EEEEFC"}
					pressColor={"#CFCFFF"}
					textColor={"#3F41FA"}
					onChange={text => this.setState({amount: text})}
					decimal={true}
					arrow={"purple"}
					value={this.state.amount}
					maxLength={7}
					noLeadingZeros={true}
				/>

				<NextButton
					title="Choose recipient"
					disabled={
						// this.state.amount.length == 0 || amountOverBalance
						false
					}
					onPress={() => {
						this.props.enterAmount(this.state.activeCurrency, parseFloat(this.state.amount))
						this.props.navigation.navigate("SendTo")
					}}/>

				<CloseButton
					color="primary"
					onPress={() => {
						this.props.screenProps.rootNavigation.goBack(null);
					}}
				/>
			</View>
		);
	}
}

const styles = StyleSheet.create({
	wrapper: {
		flex: 1,
		paddingTop: isIphoneX() ? 60 : 40,
		backgroundColor: colors.white,
		flexDirection: "column",
		justifyContent: "space-between",
		paddingBottom: isIphoneX() ? 140 : 120,
	},
	header: {
		flexDirection: "column",
		justifyContent: "center",
		alignItems: "center",
	},
	title: {
		color: colors.darkGray,
		fontSize: 24,
		fontWeight: "600",
		marginBottom: 4
	},
	balance: {
		fontSize: 18,
	},
	bodyWrapper: {
		flexDirection: "row",
		justifyContent: "center",
		alignItems: "center",
		padding: 20,
		// backgroundColor: colors.gray,
		position: "relative",
	},
	amountWrapper: {
		// backgroundColor: colors.gray,
		flexDirection: "column",
		justifyContent: "center",
		alignItems: "center",
	},
	amountTextWrapper: {
		flexDirection: "row",
		justifyContent: "center",
		alignItems: "center",
		height: 80,
	},
	amount: {
		fontWeight: "700",
		fontSize: 54,
	},
	amountSuffix: {
		fontWeight: "700",
		paddingBottom: 15,
	},
});

export default EnterAmount;
