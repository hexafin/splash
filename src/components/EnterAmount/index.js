import EnterAmount from "./EnterAmount"
import {connect} from "react-redux"
import {bindActionCreators} from "redux"
import {setActiveCurrency} from "../../redux/crypto/actions"
import {enterAmount} from "../../redux/payFlow"

const mapStateToProps = (state) => {
    return {
    	activeCurrency: state.crypto.activeCurrency,
    	balance: state.crypto.balance,
    	exchangeRates: state.crypto.exchangeRates
    }
}

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
    	setActiveCurrency,
    	enterAmount,
    }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(EnterAmount)
