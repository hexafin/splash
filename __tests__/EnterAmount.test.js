import "react-native";
import React from "react";
import EnterAmount from "../src/components/EnterAmount/EnterAmount";

it("Enter Amount renders correctly", () => {
	const snap = shallow(
		<EnterAmount
			balance={{
				BTC: 100
			}}
			activeCryptoCurrency={"BTC"}
			exchangeRates={{
				BTC: {
					USD: 1000
				}
			}}
		/>
	);
	expect(snap).toMatchSnapshot();
});
