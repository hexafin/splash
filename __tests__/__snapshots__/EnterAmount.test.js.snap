// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enter Amount renders correctly 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-between",
      "paddingBottom": 140,
      "paddingTop": 60,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <Component
      style={
        Object {
          "color": undefined,
          "fontSize": 24,
          "fontWeight": "600",
          "marginBottom": 4,
        }
      }
    >
      Sending 
      Bitcoin
    </Component>
  </Component>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "padding": 20,
        "position": "relative",
      }
    }
  >
    <CurrencySwitcherLight
      activeCurrency="USD"
      crypto="BTC"
      fiat="USD"
      onSwitch={[Function]}
      style={
        Object {
          "left": 20,
          "position": "absolute",
          "top": 40,
        }
      }
      switcherColor="purple"
      textColor="#4547E0"
      textSize={16}
    />
    <Component
      pointerEvents="none"
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <AnimatedComponent
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 80,
              "justifyContent": "center",
            },
            Object {
              "transform": Array [
                Object {
                  "scale": 1,
                },
              ],
            },
          ]
        }
      >
        <Component
          style={
            Array [
              Object {
                "fontWeight": "700",
                "paddingBottom": 15,
              },
              Object {
                "color": "#4547E0",
                "fontSize": 34,
                "paddingRight": 3,
              },
            ]
          }
        >
          $
        </Component>
        <Component
          style={
            Array [
              Object {
                "fontSize": 54,
                "fontWeight": "700",
              },
              Object {
                "color": "#4547E0",
                "opacity": 0.8,
              },
            ]
          }
        >
          0
        </Component>
      </AnimatedComponent>
      <Component
        style={
          Array [
            Object {
              "fontSize": 18,
            },
            Object {
              "color": "#919191",
            },
          ]
        }
      >
        Balance: 
        $
        100000.00
      </Component>
    </Component>
  </Component>
  <Keypad
    arrow="purple"
    decimal={true}
    delete={true}
    onChange={[Function]}
    pressColor="#CFCFFF"
    primaryColor="#EEEEFC"
    textColor="#3F41FA"
    value=""
  />
  <NextButton
    disabled={true}
    onPress={[Function]}
    title="Choose recipient"
  />
  <CloseButton
    color="dark"
    onPress={[Function]}
  />
</Component>
`;
